### PR TITLE
Rust/Swift: Make all public AST classes `final`

### PR DESCRIPTION
**Pull request by hvitved**
This PR changes the code generator used by Rust and Swift so that all public classes are `final`. Commit-by-commit review is strongly recommended.

### Before this PR

An entry such as `Expr` in `schema.py` would give rise to two classes:

**`generated/Expr.qll`**
```ql
// generated, do not edit
module Generated {
   class Expr extends Synth::TExpr, AstNode { }
}
```

**`elements/Expr.qll`**
```ql
// generated, remove this comment if you wish to edit this file
class Expr extends Generated::Expr { }
```

Any customizations needed on `Expr`, such as the addition of new predicates or overriding of existing predicates, would be done in `elements/Expr.qll` and this would also be the public version of `Expr`.

### After this PR

We want to avoid unintended predicate dispatch, i.e. make the public `Expr` class `final`. However, we cannot simply make the `Expr` class in `elements/Expr.qll` `final`, because other customization classes may want to extend it. This PR therefore adds an additional layer:

**`generated/Expr.qll`**
```ql
// generated, do not edit
module Generated {
   class Expr extends Synth::TExpr, Impl::AstNode { } // note the `Impl::` prefix
}
```

**`elements/ExprImpl.qll`**
```ql
// generated, remove this comment if you wish to edit this file
module Impl {
  class Expr extends Generated::Expr { }
}
```

**`elements/Expr.qll`**
```ql
// generated, do not edit
final class Expr = Impl::Expr;
```

Now the exposed `Expr` class is `final`, and all customizations are done in `elements/ExprImpl.qll`.

### CODE REVIEW

**Comment by redsun82** [comment]
> ```ql
> // generated, remove this comment if you wish to edit this file
> module {
>   class Expr extends Generated::Expr { }
> }
> ```

meta-review comment: a `module Generated` is missing in the PR description there :laughing:

File: misc/codegen/lib/ql.py [hunk 1]
```diff
@@ -115,6 +116,8 @@ class Class:
 
     def __post_init__(self):
         self.bases = [Base(str(b), str(prev)) for b, prev in zip(self.bases, itertools.chain([""], self.bases))]
+        self.bases_impl = [Base(str(b), str(prev)) for b, prev in zip(
+            self.bases_impl, itertools.chain([""], self.bases_impl))]
```
Comment (redsun82) [inline misc/codegen/lib/ql.py:120]: nit, there's some duplication here. Maybe
```suggestion
        get_bases = lambda bases: [Base(str(b), str(prev)) for b, prev in zip(bases, itertools.chain([""], bases))]
        self.bases = get_bases(self.bases)
        self.bases_impl = get_bases(self.bases_impl)
```
?

**Comment by hvitved** [comment]
> meta-review comment: a `module Generated` is missing in the PR description there 😆

Thanks; actually it should be `module Impl`.

**Review by redsun82 — DISMISSED** [review]
I really like this, thanks for the hard work!

I think we should move all stubs to a separate directory (`impl` or `internal`), but I saw on slack you agree with this for a follow-up PR, so I'm perfectly happy with that.

**Comment by redsun82** [comment]
ah, that pesky autopep again :sweat_smile: 

Ah, forgot to mention, the Doc coverage failure is pretty normal, because the baseline situation (with uncovered defs) gets lost in the renames. I'm ok with ignoring it.

**Comment by hvitved** [comment]
> Ah, forgot to mention, the Doc coverage failure is pretty normal, because the baseline situation (with uncovered defs) gets lost in the renames. I'm ok with ignoring it.

Yes, it was also my impression that this was because of the renamed files.

**Review by redsun82 — APPROVED** [review]

**Comment by geoffw0** [comment]
You've checked in a merge artifact `CommandInjection.expected.orig`.  Could you do a follow-up to remove it please (I appreciate it's easy to miss stuff like this with so many other files touched by this PR).

Also why is no change note required for Swift - are we confident users can't or have been told they shouldn't be overriding these classes?

**Comment by hvitved** [comment]
> You've checked in a merge artifact `CommandInjection.expected.orig`.  Could you do a follow-up to remove it please (I appreciate it's easy to miss stuff like this with so many other files touched by this PR).
> 
Yes, sorry, thanks for noticing.
> 
> Also why is no change note required for Swift - are we confident users can't or have been told they shouldn't be overriding these classes?

I'll add a change note.

**Pull request merged**